### PR TITLE
feat(scratch deploy): mvp of local scratch deployment

### PR DESCRIPTION
**Pull request by arwer13**
Renovation of the scratch deploy scripts. For the README see docs/scratch-deploy.md

NB that the scripts are to undertake one more refactoring iteration during migration to TypeScript and new repo https://github.com/lidofinance/core

### CODE REVIEW

File: SCRATCH_DEPLOY.md [hunk 1]
```diff
@@ -1,84 +1,35 @@
...
-**NB**: at the moment the deployment from scratch scripts and manual target Lido V1 version (before Lido V2 upgrade) and is not working
-
-Video guide: [youtube](https://www.youtube.com/watch?v=dCMXcfglJv0)
-
 ## Requirements
 
-* shell - bash or zsh
-* docker
-* node.js v14
+* node.js v16
```
Comment (TheDZhon) [inline SCRATCH_DEPLOY.md:5]: This version is EOL already, mind switching to v18/v20?

File: deployed-goerli.json
[the remark this answers is not in the document]
Comment (arwer13) [reply deployed-goerli.json:1]: I'm pro for the unified order and for the alphabetical one: seems to be the most seamless among other options.

**Review by loga4 — COMMENTED** [review]
maybe remove `cli` folder at all?

File: scripts/helpers/deploy.js [hunk 2]
```diff
@@ -193,91 +197,108 @@ async function deployWithoutProxy(nameInState, artifactName, deployer, construct
...
+
+  console.log(`done: ${contract.address} (gas used ${gasUsed})`)
+
+  state[nameInState] = { ...state[nameInState] }
+  state[nameInState].implementation = {
+    contract: await getContractPath(artifactName),
+    address: contract.address,
+    constructorArgs: constructorArgs,
+  }
+  persistNetworkState2(network.name, netId, state)
```
Comment (krogla) [inline scripts/helpers/deploy.js:238]: persistNetworkState2 -> persistNetworkState ?

**Review by krogla — APPROVED** [review]
couple of minor remarks. otherwise lgtm

File: scripts/helpers/deploy.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-const { readNetworkState, persistNetworkState2 } = require('./persisted-network-state')
+const { readNetworkState, persistNetworkState, persistNetworkState2 } = require('./persisted-network-state')
```
[the remark this answers is not in the document]
Comment (krogla) [reply scripts/helpers/deploy.js:1]: actually, we don't )
persistNetworkState2 was introduced by me once upon a time to maintain compatibility with all other existing scripts, but I also agree that it is a bit redundant

File: contracts/common/interfaces/IGateSealFactory.sol [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// SPDX-FileCopyrightText: 2023 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+// See contracts/COMPILERS.md
+// solhint-disable-next-line
+pragma solidity >=0.4.24 <0.9.0;
+
+interface IGateSealFactory {
```
Comment (TheDZhon) [inline contracts/common/interfaces/IGateSealFactory.sol:8]: shall we add a link to https://github.com/lidofinance/gate-seals/blob/main/contracts/GateSealFactory.vy?

File: deployed-goerli.json
[the remark this answers is not in the document]
Comment (TheDZhon) [reply deployed-goerli.json:1]: I support the suggestion by @avsetsin :+1:

File: docs/scratch-deploy.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  - It gets filled with the deployed contracts info from step to step.
+- (optional) Deploy DepositContract.
+  - The step is skipped if the DepositContract address is specified
+- (optional) Deploy ENS
+  - The step is skipped if the ENS Registry address is specified
+- Deploy Aragon framework environment
+- Deploy standard Aragon apps contracts (like `Agent`, `Voting`)
+- Deploy `LidoTemplate` contract
+  - This is an auxiliary deploy contract, which performs DAO configuration
+- Deploy Lido custom Aragon apps implementations (aka bases), namely for `Lido`, `LegacyOracle`, `NodeOperatorsRegistry`)
```
Comment (TheDZhon) [inline docs/scratch-deploy.md:41]: extra ')'

File: docs/scratch-deploy.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+To run hardhat node execute:
+```shell
+yarn hardhat node
+```
+
+## Holešky deployment
+
+To do Holešky deployment, the following parameters must be set up via env variables:
+
+- `DEPLOYER`. The deployer address, you must have its private key. It must have enough ether.
```
Comment (TheDZhon) [inline docs/scratch-deploy.md:91]: how much is enough tho?

File: docs/scratch-deploy.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+- `RPC_URL`. Address of of the Ethereum RPC node to use. E.g. for Infura it is `https://holesky.infura.io/v3/<yourProjectId>`
+- `GAS_PRIORITY_FEE`. Gas priority fee. By default set to `2`
+- `GAS_MAX_FEE`. Gas max fee. By default set to `100`
+- `GATE_SEAL_FACTORY`. Address of the [GateSeal Factory](https://github.com/lidofinance/gate-seals) contract. Must be deployed preliminary. Can be set to any `0x0000000000000000000000000000000000000000` to debug deployment.
```
Comment (TheDZhon) [inline docs/scratch-deploy.md:95]: maybe not 'preliminary' but 'beforehand' or 'in advance'

File: docs/scratch-deploy.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+NB, that some contracts require additional auxiliary contract to be deployed. Namely, constructor of `AppProxyPinned` depends on proxy implementation ("base" in Aragon terms) contract with `initialize()` function and `Kernel` contract which must return the implementation by call `kernel().getApp(KERNEL_APP_BASES_NAMESPACE, _appId)`. See `@aragon/os/contracts/apps/AppProxyBase.sol` for the details.
+
+## Post deploy initialization
+
+### Initialization up to fully operational state
+
+In order to make protocol fully operational the additional steps are required.
+
+- add `NodeOperatorsRegistry` as staking module: `StakingRouter.addStakingModule`
```
Comment (TheDZhon) [inline docs/scratch-deploy.md:134]: it was done already AFAIK bu the script, see the Line 52:
> - Plug NodeOperatorsRegistry as Curated staking module

File: docs/scratch-deploy.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+- add oracle committee members to `HashConsensus` contracts for `AccountingOracle` and `ValidatorsExitBusOracle`: `HashConsensus.addMember`
+- initialize initial epoch for `HashConsensus` contracts for `AccountingOracle` and `ValidatorsExitBusOracle`: `HashConsensus.updateInitialEpoch`
+- add guardians to `DepositSecurityModule`: `DepositSecurityModule.addGuardians`
+- resume protocol: `Lido.resume`
+- resume WithdrawalQueue: `WithdrawalQueueERC721.resume`
+- add at least one Node Operator: `NodeOperatorsRegistry.addNodeOperator`
+- add validator keys to the Node Operators: `NodeOperatorsRegistry.addSigningKeys`
+- set staking limits for the Node Operators: `NodeOperatorsRegistry.setNodeOperatorStakingLimit`
+
+NB, that part of the actions require preliminary granting of the required roles, e.g. `STAKING_MODULE_MANAGE_ROLE` for `StakingRouter.addStakingModule`:
```
Comment (TheDZhon) [inline docs/scratch-deploy.md:144]: not 'preliminary' maybe (see above)

File: docs/scratch-deploy.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    nodeOperatorsRegistry.address,
+    NOR_STAKING_MODULE_TARGET_SHARE_BP,
+    NOR_STAKING_MODULE_MODULE_FEE_BP,
+    NOR_STAKING_MODULE_TREASURY_FEE_BP,
+    { from: agent.address }
+  )
+  await stakingRouter.renounceRole(STAKING_MODULE_MANAGE_ROLE, agent.address, { from: agent.address })
+```
+
+## Protocol parameters
```
Comment (TheDZhon) [inline docs/scratch-deploy.md:159]: Have you checked the final ACL setup against the https://docs.lido.fi/guides/verify-lido-v2-upgrade-manual#roles-setup ? (esp. Burner-related roles for NOR and Lido)

File: hardhat.config.js [hunk 11]
```diff
@@ -44,69 +51,57 @@ const getNetConfig = (networkName, ethAccountName) => {
...
+      // NB!: forking get enabled if env variable HARDHAT_FORKING_URL is set, see code below
       blockGasLimit: 30000000,
       gasPrice: 0,
       initialBaseFeePerGas: 0,
       allowUnlimitedContractSize: true,
       accounts: {
         // default hardhat's node mnemonic
         mnemonic: 'test test test test test test test test test test test junk',
-        count: 30,
+        count: 10,
```
Comment (TheDZhon) [inline hardhat.config.js:78]: :warning:  I am not sure that it wouldn't break some tests which depend on multiple accounts, actually 
e.g., sanity checker tests if I am not wrong

File: hardhat.config.js [hunk 12]
```diff
@@ -115,12 +110,15 @@ const getNetConfig = (networkName, ethAccountName) => {
       chainId: 1,
       timeout: 60000 * 10,
       forking: {
-        url: 'https://mainnet.infura.io/v3/' + accounts.infura.projectId,
+        url: RPC_URL,
         // url: 'https://eth-mainnet.alchemyapi.io/v2/' + accounts.alchemy.apiKey
```
Comment (TheDZhon) [inline hardhat.config.js:114]: inconsistent comments (infura / alchemy, whatever)

File: scripts/helpers/assert.js [hunk 13]
```diff
@@ -13,6 +13,10 @@ assert.addressEqual = (actual, expected, msg) => {
   assert.equal(toChecksumAddress(actual), toChecksumAddress(expected), msg)
 }
 
+assert.hexEqual = (actual, expected, msg) => {
```
Comment (TheDZhon) [inline scripts/helpers/assert.js:16]: The name suggests that the data is hex, should we check for the allowed symbols then?
e.g., smth like `0[xX][0-9a-fA-F]+`

File: scripts/helpers/deploy.js [hunk 14]
```diff
@@ -193,91 +197,108 @@ async function deployWithoutProxy(nameInState, artifactName, deployer, construct
...
-    console.log(`Constructor args: ${constructorArgs}`)
-  }
+  process.stdout.write(`Deploying ${artifactName} (NO proxy)... `)
 
   const contract = await deployContract(artifactName, constructorArgs, deployer)
-  console.log(`${artifactName} (NO proxy): ${contract.address}`)
 
   const gasUsed = await getDeploymentGasUsed(contract)
-  console.log(`    gas used ${gasUsed}`)
+  console.log(`done: ${contract.address} (gas used ${gasUsed})`)
```
Comment (TheDZhon) [inline scripts/helpers/deploy.js:205]: :brain:  I'd suggest to have the logs not only printed but dumped to the file, not a strong opinion tho

File: scripts/helpers/deploy.js [hunk 15]
```diff
@@ -193,91 +197,108 @@ async function deployWithoutProxy(nameInState, artifactName, deployer, construct
...
+}
+
 async function deployBehindOssifiableProxy(nameInState, artifactName, proxyOwner, deployer, constructorArgs=[], implementation=null) {
   const netId = await web3.eth.net.getId()
   const state = readNetworkState(network.name, netId)
+  const proxyContractName = 'OssifiableProxy'
 
   if (implementation === null) {
-    console.log(`Deploying implementation for proxy of ${artifactName}`)
+    process.stdout.write(`Deploying implementation for proxy of ${artifactName}... `)
```
Comment (TheDZhon) [inline scripts/helpers/deploy.js:248]: why sometimes `console.log` but here `process.stdout`?

File: scripts/helpers/run-or-wrap-script.js [hunk 16]
```diff
@@ -15,7 +15,7 @@ module.exports = (scriptFn, mainModule) => {
     // Buidler executes scripts in a forked subprocess
     scriptFn({ artifacts: globalArtifacts, web3: globalWeb3 })
       .then(() => {
-        console.error('All done!')
+        console.error('')
```
Comment (TheDZhon) [inline scripts/helpers/run-or-wrap-script.js:18]: looks a bit weird

File: scripts/scratch/00-populate-deploy-artifact-from-env.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
+const { ZERO_ADDRESS } = require('@aragon/contract-helpers-test')
+const runOrWrapScript = require('../helpers/run-or-wrap-script')
+const { readNetworkState, persistNetworkState } = require('../helpers/persisted-network-state')
+
+const DEPLOYER = process.env.DEPLOYER
+const CHAIN_ID = parseInt(process.env.CHAIN_ID)
```
Comment (TheDZhon) [inline scripts/scratch/00-populate-deploy-artifact-from-env.js:6]: Actually, we can check the provided `CHAIN_ID` against the real one using the `eth_chainid` RPC call

File: scripts/scratch/checks/scratch-acceptance-test.js [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+  assert.equals(await lido.getTotalPooledEther(), initialLidoBalance + BigInt(ETH(34)))
+  assert.equals(await lido.getBufferedEther(), initialLidoBalance + BigInt(ETH(34)))
+
+  await lido.deposit(MAX_DEPOSITS, CURATED_MODULE_ID, CALLDATA, { from: depositSecurityModule.address })
+  log.success('Ether deposited')
+
+
+  assert.equals((await lido.getBeaconStat()).depositedValidators, 1)
+
+  // const checkStat = async ({ depositedValidators, beaconValidators, beaconBalance }) => {
```
Comment (TheDZhon) [inline scripts/scratch/checks/scratch-acceptance-test.js:168]: let's remove the commented code

File: scripts/scratch/dao-deploy.sh [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+yarn hardhat --network $NETWORK run ./scripts/scratch/16-plug-curated-staking-module.js --no-compile
+msg "Plugged NodeOperatorsRegistry as Curated staking module"
+
+yarn hardhat --network $NETWORK run ./scripts/scratch/17-transfer-roles.js --no-compile
+msg "Role admin permissions transferred to Agent"
+
+yarn hardhat --network $NETWORK run ./scripts/scratch/12-check-dao.js --no-compile
+msg "The deployed protocol state checked"
+
+# TODO: save commit of the latest deploy
```
Comment (TheDZhon) [inline scripts/scratch/dao-deploy.sh:96]: ?

**Review by TheDZhon — CHANGES_REQUESTED** [review]
Wow, this is huge :clap: 

I'd suggest implementing some changes and updating the README file as well to properly refer to the new scratch deployment approach.

**Comment by TheDZhon** [comment]
@arwer13 could you please explicitly secure the quorum-sufficient TESTLDO amount to reside on the Agent's balance during the deployment

That would ease governance-related test fixtures greatly.

@bulbozaur proposes the idea, and I'd suggest implementing it

File: scripts/scratch/dao-deploy.sh (hunk 19 above)
Comment (arwer13) [reply scripts/scratch/dao-deploy.sh:96]: just removed it

File: scripts/scratch/00-populate-deploy-artifact-from-env.js (hunk 17 above)
Comment (arwer13) [reply scripts/scratch/00-populate-deploy-artifact-from-env.js:6]: I've to conclusion it is excess. Can take if from RPC call to record in the state file. The checks for correspondence of the chainId will be done automatically anyway for live networks for which chainId is specified in hardhat.config.js

File: scripts/helpers/deploy.js (hunk 15 above)
Comment (arwer13) [reply scripts/helpers/deploy.js:247]: `process.stdout.write` is for printing without end of line (just to save some space on the screen)

File: scripts/helpers/deploy.js (hunk 2 above)
Comment (arwer13) [reply scripts/helpers/deploy.js]: removed duplicate function, left the more straightforward version of it

File: scripts/helpers/deploy.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (arwer13) [reply scripts/helpers/deploy.js]: removed duplicate function, left the more straightforward version of it

File: deployed-goerli.json
[the remark this answers is not in the document]
Comment (arwer13) [reply deployed-goerli.json:1]: Done

File: docs/scratch-deploy.md (hunk 6 above)
Comment (arwer13) [reply docs/scratch-deploy.md]: Approx 100,000,000 gas. Added the info to scratch-deploy.md

**Comment by arwer13** [comment]
> could you please explicitly secure the quorum-sufficient TESTLDO amount to reside on the Agent's balance during the deployment
Added.

The other comments are processed as well.
Ready for merge, imho.

**Pull request merged**